### PR TITLE
opt(lua): Avoid separate hops for lock & unlock on single shard execution

### PR DESCRIPTION
**Pull request by chakaz**

### CODE REVIEW

**Review by dranikpg — APPROVED** [review]
I'm not against approving

But what I thought is that we can do without the implementation. Just make the multi transaction non multi and a run a single hop on it. 

So this pr can be collapsed down to the ifs in EvalInternal and a single multi_.reset()

**Comment by chakaz** [comment]
I can definitely give this idea a shot (not right now, but later today when I'm back at work), but I would say that in my previous attempts around this code, without multi, I had to do many tweaks due to subtle reentrant issues (SingleHop calling SingleHop, etc).
The squashing hack that you came up with only works for multi (currently), do you mean adapting it to work without multi?

**Comment by dranikpg** [comment]
You use squashing, so there is no re-entrancy. You use squashing on the strub transaction, so it doesn't matter what the original transaction is

File: src/server/main_service.cc [hunk 1]
```diff
@@ -1565,52 +1564,66 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
...
-
   interpreter->SetGlobalArray("KEYS", eval_args.keys);
   interpreter->SetGlobalArray("ARGV", eval_args.args);
+  absl::Cleanup clean = [interpreter]() { interpreter->ResetStack(); };
 
   Interpreter::RunResult result;
-  optional<ShardId> sid = GetRemoteShardToRunAt(*tx);
-  if (tx->GetMultiMode() != Transaction::NON_ATOMIC && sid.has_value()) {
+
+  if (CanRunSingleShardMulti(sid, *params, tx)) {
```
Comment (romange) [inline src/server/main_service.cc:1597]: if you pass params as const ref you can do the same with tx

File: src/server/main_service.cc [hunk 2]
```diff
@@ -1565,52 +1564,66 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
...
       CallFromScript(cntx, args);
     });
 
-    boost::intrusive_ptr<Transaction> stub_tx = new Transaction{cntx->transaction};
-    cntx->transaction = stub_tx.get();
-    tx->PrepareSquashedMultiHop(registry_.Find("EVAL"), [&](ShardId id) { return id == *sid; });
-    tx->ScheduleSingleHop([&](Transaction* tx, EngineShard*) {
+    tx->RunSingleShardMulti(cntx->db_index(), eval_args.keys, [&]() {
       ++ServerState::tlocal()->stats.eval_shardlocal_coordination_cnt;
+      boost::intrusive_ptr<Transaction> stub_tx = new Transaction{tx};
```
Comment (romange) [inline src/server/main_service.cc:1607]: i would pull out the code that does not have to be in the callback. i.e. lines 1606-1608

**Comment by romange** [comment]
+1 to Vlad's option if we make it work. The goal is to force ScheduleSingleHop to execute the same flow you wrote in RunSingleShardMulti.

**Comment by chakaz** [comment]
Indeed, I am very close to getting it to work with `ScheduleSingleHop` instead of my new hook.
It's a bit trickier than I'd hoped, but much better & simpler still

File: src/server/transaction.cc [hunk 3]
```diff
@@ -405,6 +410,15 @@ string Transaction::DebugId() const {
...
 }
 
+void Transaction::PrepareMultiForScheduleSingleHop(ShardId sid, DbIndex db, CmdArgList args) {
+  multi_.reset();
+  InitBase(db, args);
+  EnableShard(sid);
+  OpResult<KeyIndex> key_index = DetermineKeys(cid_, args);
+  CHECK(key_index);
+  StoreKeysInArgs(*key_index, false);
+}
```
Comment (dranikpg) [inline src/server/transaction.cc:420]: Looks very elegant now 😎

File: src/server/main_service.cc [hunk 4]
```diff
@@ -1565,52 +1564,70 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
...
+
+    ShardId cur_sid = Shard(key, shard_count());
+    if (i == 0) {
+      sid = cur_sid;
+    }
+    if (sid.has_value() && *sid != cur_sid) {
+      sid = nullopt;
+    }
   }
+
```
Comment (dranikpg) [inline src/server/main_service.cc:1586]: The only thing I'm worried about is that we compute hashes twice now. I remember when I was benchmarking tx code a few months ago, computing Shard() actually took up to 5% of execution time.

What we could do in the _future_, is for example avoid Scheudling in `StartMultiLockedAhead`. And then also re-use this function for MULTI/EXEC. This way we don't need to compute hashes and re-initialize the transaction

**Review by dranikpg — APPROVED** [review]

File: src/server/main_service.cc (hunk 4 above)
Comment (chakaz) [reply src/server/main_service.cc:1586]: That's a good idea!

**Comment by chakaz** [comment]
FWIW, I just double checked that the performance of the latest iteration is on par with what I had originally sent, and indeed they are the same. As such, I'll now merge the changes :partying_face:

**Pull request merged**